sympy/stats/crv.py

```diff
@@ -531,11 +531,13 @@ def compute_density(self, expr, **kwargs):
 
         gs = solveset(expr - y, self.value, S.Reals)
 
-        if isinstance(gs, Intersection) and S.Reals in gs.args:
-            gs = list(gs.args[1])
+        if isinstance(gs, Intersection):
+            if len(gs.args) == 2 and gs.args[0] is S.Reals:
```

Comment:
Is there a test that affirms that S.Reals will always be in slot 0 of any intersection with 2 args?

Reply:
I don't think there's anything like that. This patch comes from https://github.com/sympy/sympy/issues/24915#issuecomment-1470379843, but I'm not clear why Oscar wrote the logic like that.

Reply:
The current code is:
```
        if isinstance(gs, Intersection) and S.Reals in gs.args:
            gs = list(gs.args[1])
```
The intention of this code is that we have `Intersection(Reals, S)` and we want `S` then we can check if Reals is in the args and then select `.args[1]`. This makes presumptions that are not guaranteed to be true such as that there are only two args and that if Reals is in the args then it would be as position `0` so that `S` would be at position `1`.

The changed code I showed is not a great fix but it removes those incorrect assumptions:
```
        if isinstance(gs, Intersection):
            if len(gs.args) == 2 and gs.args[0] is S.Reals:
                gs = gs.args[1]
```
Now we know that we have `Intersection(Reals, S)` and we are getting `S` out.

As for a better fix I think that this code should not attempt to peel off an intersection but this just shows how poor solveset's interface is for some of the cases where it might be suggested to use it. The proper fix should really be to not use solveset at all here because if the output needs to be manipulated like this then solveset is not the right function to use.

This code clearly wants explicit solution expressions and solve is the function that provides those. There are various problems with solve but solveset does not in any way function as a replacement for it. The usage of solveset here is inappropriate and most likely dates back to the previous push to "replace solve with solveset" which was a mistake that should be reversed: solveset is not an appropriate replacement for most uses of solve.